chips/lowrisc/src/flash_ctrl.rs

```diff
@@ -0,0 +1,553 @@
+                }
+                self.enable_interrupts();
+            });
+        }
+
+        if irqs.is_set(INTR::OP_DONE) {
+            let read_buf = self.read_buf.take();
+            if let Some(buf) = read_buf {
+                // We were doing a read
+                if self.read_index.get() >= buf.0.len() {
+                    // We have all of the data, call the client
+                    self.flash_client.map(move |client| {
+                        client.read_complete(buf, hil::flash::Error::CommandComplete);
+                    });
+                } else {
+                    // Still waiting on data, keep waiting
+                    self.read_buf.replace(buf);
+                    self.enable_interrupts();
+                }
+            }
+
+            let write_buf = self.write_buf.take();
+            if let Some(buf) = write_buf {
+                // We were doing a write
+                if self.write_index.get() >= buf.0.len() {
+                    // We sent all of the data, call the client
+                    self.flash_client.map(move |client| {
+                        client.write_complete(buf, hil::flash::Error::CommandComplete);
+                    });
+                } else {
+                    // Still writing data, keep trying
+                    self.write_buf.replace(buf);
+                    self.enable_interrupts();
+                }
+            }
+        }
+    }
+}
+
+impl<C: hil::flash::Client<Self>> hil::flash::HasClient<'static, C> for FlashCtrl<'_> {
+    fn set_client(&self, client: &'static C) {
+        self.flash_client.set(client);
+    }
+}
+
+impl hil::flash::Flash for FlashCtrl<'_> {
+    type Page = LowRiscPage;
+
+    fn read_page(
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> Result<(), (ReturnCode, &'static mut Self::Page)> {
+        let addr = page_number * PAGE_SIZE;
+
+        if !self.data_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_data_partition(self.region_num);
+        }
+
+        if !self.info_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_info_partition(FlashBank::BANK0, self.region_num);
+        }
+
+        // Enable interrupts and set the FIFO level
+        self.enable_interrupts();
+        self.registers.fifo_lvl.modify(FIFO_LVL::RD.val(0xF));
+
+        // Set the address
+        self.registers.addr.write(ADDR::START.val(addr as u32));
+
+        // Save the buffer
+        self.read_buf.replace(buf);
+        self.read_index.set(0);
+
+        // Start the transaction
+        let info_sel = if self.region_num as u32 > 3 {
+            CONTROL::INFO_SEL::SET
+        } else {
+            CONTROL::INFO_SEL::CLEAR
+        };
+        self.registers.control.write(
+            CONTROL::OP::READ
+                + CONTROL::PARTITION_SEL::INFO
```

Comment:
If I understand the docs correctly, this means that all flash reads and writes through this interface are exclusively being made to the info partition. However, the docs also state "The data partition holds generic data like a generic memory would. The information partition holds metadata about the data partition as well design specific secret data", which to me suggests that for the generic flash interface exposed by Tock we should be conducting all reads and writes in the data partition, rather than an info partition.

Am I missing something, or should this change here and in `write_page()` / `erase_page()` to select the data partition?

Reply:
The docs says that

> data partition - this is the portion of flash that is accessible both by the host and by the controller.

So I don't think data partition is the right place to store sensitive data as then the host can read it.

Reply:
What is the host in this case? Does that just mean that the CPU can directly perform memory mapped reads from flash?

I thought that is how the flash works on most chips in Tock -- the nonvolatile storage interface is not assumed to be used for secrets, but just exposes a way to read/write to allowed locations in flash memory. Generally, we rely on `#[forbid_unsafe]` to prevent capsules from accessing arbitrary flash memory for reads using memory accesses.

But I am not the most familiar with how the nonvolatile storage interface is intended to work, so I will defer to others on this.

Reply:
In the case of EarlGrey, I believe the host is Ibex.

Reply:
Ah, I was under the impression that host meant the CPU OpenTitan was connected to.

Ok, so it sounds like the data partition would be a better place to use for Tock's flash.